staging/src/k8s.io/apiserver/pkg/reconcilers/peer_endpoint_lease.go

```diff
@@ -0,0 +1,312 @@
+type PeerLeaseReconciler interface {
+	// ListLeases retrieves a list of the current server IPs
+	ListLeases() ([]string, error)
+
+	// UpdateLease adds or refreshes a servers lease
+	UpdateLease(ip string, peerBindAddress string, serverId string, endpointPorts []corev1.EndpointPort) error
+
+	// GetLease retrieves the IP of a specific service
+	GetLease(serverId string) (string, error)
+
+	// RemoveLease removes a server's lease
+	RemoveLease(ip string, endpointPorts []corev1.EndpointPort) error
+
+	// Destroy cleans up everything on shutdown.
+	Destroy()
+}
+
+type PeerEndpointLeases struct {
+	Storage   storage.Interface
+	DestroyFn func()
+	BaseKey   string
+	LeaseTime time.Duration
+}
+
+// ListLeases retrieves a list of the current server IPs from storage
+func (s *PeerEndpointLeases) ListLeases() ([]string, error) {
+	storageOpts := storage.ListOptions{
+		ResourceVersion:      "0",
+		ResourceVersionMatch: metav1.ResourceVersionMatchNotOlderThan,
+		Predicate:            storage.Everything,
+		Recursive:            true,
+	}
+	ipInfoList, err := s.GetIpInfoList(storageOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	ipList := make([]string, 0, len(ipInfoList.Items))
+	for _, ip := range ipInfoList.Items {
+		if len(ip.Subsets) > 0 && len(ip.Subsets[0].Addresses) > 0 && len(ip.Subsets[0].Addresses[0].IP) > 0 {
+			ipList = append(ipList, ip.Subsets[0].Addresses[0].IP)
+		}
+	}
+
+	klog.V(6).Infof("Current server IPs listed in storage are %v", ipList)
+
+	return ipList, nil
+}
+
+// UpdateLease resets the TTL on a server IP in storage
+// UpdateLease will create a new key if it doesn't exist.
+// We use the first element in endpointPorts as a part of the lease's base key
+// This is done to support out tests that simulate 2 apiservers running on the same ip but
+// different ports
+
+// It will also do the following if UnknownVersionInteroperabilityProxy feature is enabled
+// 1. store the apiserverId as a label
+// 2. store the values passed to --peer-bind-ip and --peer-bind-port flags to kube-apiserver as an annotation
+// with value of format <ip:port>
+func (s *PeerEndpointLeases) UpdateLease(ip string, peerBindAddress string, serverId string, endpointPorts []corev1.EndpointPort) error {
+	key := path.Join(s.BaseKey, ip)
+	if len(endpointPorts) > 0 {
+		key = path.Join(key, fmt.Sprint(endpointPorts[0].Port))
+	}
+	return s.Storage.GuaranteedUpdate(apirequest.NewDefaultContext(), key, &corev1.Endpoints{}, true, nil, func(input kruntime.Object, respMeta storage.ResponseMeta) (kruntime.Object, *uint64, error) {
+		existing := input.(*corev1.Endpoints)
+		existing.Subsets = []corev1.EndpointSubset{
+			{
+				Addresses: []corev1.EndpointAddress{{IP: ip}},
+				Ports:     endpointPorts,
+			},
+		}
+
+		// store this server's identity (serverId) as a label. This will be used by
+		// peers to find the IP of this server when the peer can not serve a request
+		// due to version skew.
+		if serverId != "" && PeerProxyFeatureEnabled() {
```

Comment:
This check is actually redundant since we are in fact checking for the feature flag much earlier [here](https://github.com/kubernetes/kubernetes/pull/117740/files#diff-05cd394e18da9e91dbf8b3d82e1e0813be670ee37db43ca5af6f91a39f78d08fR498-R520)

Reply:
Let's remove this check then.

Reply:
yes removed.